### PR TITLE
Implement interoperability for compressed buffers between platforms of different endianness

### PR DESCRIPTION
**Pull request by ltrk2**
The modification ensures that `CompressedWriteBuffer` and `CompressedReadBufferBase` both treat checksums as little-endian in order to provide interoperability between platforms of different endianness. It's a breaking change for big-endian platforms, but it shouldn't have any impact on little-endian ones.

This fixes functional test suite `tests/queries/0_stateless/01601_proxy_protocol.sh`.

Examples of use-cases:
* client and server are running on platforms of different endianness,
* working directory of a server is transferred among platforms of different endianness.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement interoperability between platforms for `CompressedWriteBuffer` and `CompressedReadBufferBase` by always using little-endian to store checksums.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
@ltrk2, there is no production support for ClickHouse on big-endian (it never worked), so it cannot be counted as a backward incompatible change.

File: src/Compression/Utilities.h [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include <city.h>
+
+#include <span>
+
+namespace DB
+{
+CityHash_v1_0_2::uint128 CalculateCityHash128InLittleEndian(std::span<char> buffer);
```
Comment (alexey-milovidov) [inline src/Compression/Utilities.h:9]: Missing a comment.

The function name is incorrect.
It does not calculate something "in little-endian".

Poor separation of concepts in the code.
You want to write (serialize) the components of the hash in little-endian order. So, let's write the code in another way.

I can make it.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: src/Compression/Utilities.cpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#include "Utilities.h"
+
+namespace DB
+{
+CityHash_v1_0_2::uint128 CalculateCityHash128InLittleEndian(const std::span<char> buffer)
+{
+    const auto hash = CityHash_v1_0_2::CityHash128(buffer.data(), buffer.size());
+
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
```
Comment (alexey-milovidov) [inline src/Compression/Utilities.cpp:9]: This is inconsistent with the other parts of the code base.
See `std::endian`

File: src/Compression/CompressedReadBufferBase.cpp [hunk 3]
```diff
@@ -1,4 +1,5 @@
 #include "CompressedReadBufferBase.h"
+#include "Utilities.h"
```
Comment (alexey-milovidov) [inline src/Compression/CompressedReadBufferBase.cpp:2]: Utilities is a bad name for a header file.
It means, "I want to put random unclassified garbage there".

File: src/Compression/Utilities.cpp (hunk 2 above)
Comment (ltrk2) [reply src/Compression/Utilities.cpp:9]: I would've preferred to do it this way:
```
if constexpr (std::endian::native == std::endian::big)
```
but this approach results in a build error (`error: code will never be executed [-Werror,-Wunreachable-code]`). Disabling the warnings for this scope would also have been an option, but that's also horrible. What would you recommend?

File: src/Compression/CompressedReadBufferBase.cpp (hunk 3 above)
Comment (ltrk2) [reply src/Compression/CompressedReadBufferBase.cpp:2]: I get where you're coming from, I'll come up with a proper name.

File: src/Compression/Utilities.h (hunk 1 above)
Comment (ltrk2) [reply src/Compression/Utilities.h:9]: I get your point around "in little-endian" being debatable, but it does indeed produce something in little-endian representation.

I also understand your concern that this has to do more with serialization than anything else and you don't wish to dilute other parts of the code with it.

Don't worry about it, I'll do it tomorrow.

**Comment by alexey-milovidov** [comment]
See #49005

**Pull request closed without merging**